sqlitepool/sqlitepool.go

```diff
@@ -0,0 +1,318 @@
+// One is reserved for writable transactions, the others are
+// used for read-only transactions.
+type Pool struct {
+	poolSize    int
+	rwConnFree  chan *conn // cap == 1
+	roConnsFree chan *conn // cap == poolSize-1
+	tracer      sqliteh.Tracer
+	closed      chan struct{}
+}
+
+type conn struct {
+	pool  *Pool
+	db    sqliteh.DB
+	stmts map[string]sqliteh.Stmt // persistent statements on db
+	id    sqliteh.TraceConnID
+}
+
+// NewPool creates a Pool of poolSize database connections.
+//
+// For each connection, initFn is called to initialize the connection.
+// Tracer is used to report statistics about the use of the Pool.
+func NewPool(filename string, poolSize int, initFn func(sqliteh.DB) error, tracer sqliteh.Tracer) (p *Pool, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("sqlitepool.NewPool: %w", err)
+			select {
+			case conn := <-p.rwConnFree:
+				conn.db.Close()
+			default:
+			}
+			for conn := range p.roConnsFree {
+				conn.db.Close()
+			}
+		}
+	}()
+	if poolSize < 2 {
+		return nil, fmt.Errorf("poolSize=%d is too small", poolSize)
+	}
+	p = &Pool{
+		poolSize:    poolSize,
+		rwConnFree:  make(chan *conn, 1),
+		roConnsFree: make(chan *conn, poolSize-1),
+		tracer:      tracer,
+		closed:      make(chan struct{}),
+	}
+	for i := 0; i < poolSize; i++ {
+		db, err := cgosqlite.Open(filename, sqliteh.OpenFlagsDefault, "")
+		if err != nil {
+			return nil, err
+		}
+		if err := initFn(db); err != nil {
+			return nil, err
+		}
+		c := &conn{
+			pool:  p,
+			db:    db,
+			stmts: make(map[string]sqliteh.Stmt),
+			id:    sqliteh.TraceConnID(i),
+		}
+		if i == 0 {
+			p.rwConnFree <- c
+		} else {
+			if err := ExecScript(c.db, "PRAGMA query_only=true"); err != nil {
+				return nil, err
+			}
+			p.roConnsFree <- c
+		}
+	}
+
+	return p, nil
+}
+
+func (c *conn) close() error {
+	if c.db == nil {
+		return errors.New("sqlitepool conn already closed")
+	}
+	for _, stmt := range c.stmts {
+		stmt.Finalize()
+	}
+	c.stmts = nil
+	err := c.db.Close()
+	c.db = nil
+	return err
+}
+
+func (p *Pool) Close() error {
+	select {
+	case <-p.closed:
+		return errors.New("pool already closed")
+	default:
+	}
+	close(p.closed)
+
+	c := <-p.rwConnFree
+	err := c.close()
+
+	for i := 0; i < p.poolSize-1; i++ {
+		c := <-p.roConnsFree
+		err2 := c.close()
+		if err == nil {
+			err = err2
+		}
+	}
+	return err
+}
+
+var errPoolClosed = fmt.Errorf("%w: sqlitepool closed", context.Canceled)
+
+// BeginTx creates a writable transaction using BEGIN IMMEDIATE.
+func (p *Pool) BeginTx(ctx context.Context, why string) (*Tx, error) {
+	select {
+	case <-p.closed:
+		return nil, errPoolClosed
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case conn := <-p.rwConnFree:
+		tx := &Tx{Rx: &Rx{conn: conn, inTx: true}}
+		err := tx.Exec("BEGIN IMMEDIATE;")
+		if p.tracer != nil {
+			p.tracer.BeginTx(ctx, conn.id, why, false, err)
+		}
+		if err != nil {
+			p.rwConnFree <- conn
```

Comment:
also select on closed here?

Or I guess a return of a conn to its buffered pool can't block.

Maybe a `// can't block` comment to assure future readers